torch/lowrank.py

```diff
@@ -0,0 +1,249 @@
+              :math:`q = k + 5..10`.  If k is relatively small
+              compared to :math:`min(m, n)`, choosing :math:`q = k +
+              0..2` may be sufficient.
+
+    .. note:: To obtain repeatable results, reset the seed for the
+              pseudorandom number generator
+
+    Arguments::
+        A (Tensor): the input tensor of size :math:`(*, m, n)`
+
+        q (int): the dimension of subspace spanned by Q columns.
+
+        niter (int, optional): the number of subspace iterations to
+                               conduct; niter must be a nonnegative
+                               integer. In most cases, the default
+                               value 2 is more than enough.
+
+        M (Tensor, optional): the input tensor's mean of size
+                              :math:`(*, 1, n)`.
+
+    References::
+        - Nathan Halko, Per-Gunnar Martinsson, and Joel Tropp, Finding
+          structure with randomness: probabilistic algorithms for
+          constructing approximate matrix decompositions,
+          arXiv:0909.4061 [math.NA; math.PR], 2009 (available at
+          `arXiv <http://arxiv.org/abs/0909.4061>`_).
+
+    """
+    m, n = A.shape[-2:]
+    dtype = _utils.get_floating_dtype(A)
+    matmul = _utils.get_matmul(A)
+
+    R = _utils.uniform(low=-1.0, high=1.0, size=(n, q),
+                       dtype=dtype, device=A.device)
+
+    A_H = _utils.transjugate(A)
+    if M is None:
+        (Q, _) = matmul(A, R).qr()
+        for i in range(niter):
+            (Q, _) = matmul(A_H, Q).qr()
+            (Q, _) = matmul(A, Q).qr()
+    else:
+        matmul2 = _utils.get_matmul(M)
+        M_H = _utils.transjugate(M)
+        (Q, _) = (matmul(A, R) - matmul2(M, R)).qr()
+        for i in range(niter):
+            (Q, _) = (matmul(A_H, Q) - matmul2(M_H, Q)).qr()
+            (Q, _) = (matmul(A, Q) - matmul2(M, Q)).qr()
+
+    return Q
+
+
+def svd(A, q=6, niter=2, M=None):
+    """Return the singular value decomposition ``(U, S, V)`` of a matrix,
+    batches of matrices, or a sparse matrix :math:`A` such that
+    :math:`A \approx U diag(S) V^T`. In case :math:`M` is given, then
+    SVD is computed for the matrix :math:`A - M`.
+
+    .. note:: The implementation is based on the Algorithm 5.1 from
+              Halko et al, 2009.
+
+    .. note:: To obtain repeatable results, reset the seed for the
+              pseudorandom number generator
+
+    .. note:: The input is assumed to be a low-rank matrix.
+
+    .. note:: In general, use the full-rank SVD implementation
+              ``torch.svd`` for dense matrices due to its 10-fold
+              higher performance characteristics. The low-rank SVD
+              will be useful for huge sparse matrices that
+              ``torch.svd`` cannot handle.
+
+    Arguments::
+        A (Tensor): the input tensor of size :math:`(*, m, n)`
+
+        q (int, optional): a slightly overestimated rank of A.
```

Comment:
Before checking the default values I assumed that this q, if not provided, is guaranteed to be larger than `rank(A)`.

Reply:
We cannot know the `rank(A)` in advance and it is expected that a user estimates this from the knowledge s/he has on `A`. Also, the exact rank of `A` may not be small or be even full, one can still approximate it with a low-rank matrix.

The rank of SVD (if defined by the size of `S`), is `q`.

The number of iterations for the desired error depends on the input matrix. See Part III in Halko et al, 2009, for the error analysis, in particular, Theorem 10.5 and Corollary 10.10. While giving the definite answer here is difficult, notice that if `A` is a truly low-rank matrix, the approximation becomes be exact (up to numerical errors). The current implementation is not iterative in the sense that the number of iterations is predetermined by the parameter `niter`.